📖  Update building-a-bento-amp-extension.md (#34908)
ampproject/amphtml: docs/building-a-bento-amp-extension.md

The first step to creating a new Bento AMP component is familiarizing yourself with [AMP's contributor guidelines](https://go.amp.dev/contribute/code). Adding new components follow the [process for making a significant change](https://go.amp.dev/contribute/code#process-for-significant-changes). This process includes filing an ["Intent to Implement" issue](https://go.amp.dev/i2i) and working with an assigned AMP developer guide before starting significant work.

> If the contribution of a Bento component involves a conversion of an existing component with a version bump, and is expected to span multiple PRs, it is recommended to track progress with a [Bento tracking issue](https://github.com/ampproject/amphtml/issues/new?assignees=&labels=WG%3A+bento&template=bento-component-tracker.yml&title=%F0%9F%8D%B1+%5Bamp-component-name%3Aversion%5D+Bento+tracking+issue).

Reviewer: ```suggestion
If the contribution of a Bento component involves porting an existing component to a new version, and is expected to span multiple PRs, it is recommended to track progress with a [Bento tracking issue](https://github.com/ampproject/amphtml/issues/new?assignees=&labels=WG%3A+bento&template=bento-component-tracker.yml&title=%F0%9F%8D%B1+%5Bamp-component-name%3Aversion%5D+Bento+tracking+issue).
```

## Extend AMP.PreactBaseElement

> Most AMP components extend `AMP.BaseElement`. `AMP.BaseElement` provides hookups and callbacks needed to implement and customize behavior. Callback explanations are in the BaseElement Callbacks section of [Building an AMP Extension](https://github.com/ampproject/amphtml/blob/main/docs/building-an-amp-extension.md), and inlined in the [BaseElement](https://github.com/ampproject/amphtml/blob/main/src/base-element.js) class.

Reviewer: Could we forgo a reference to methods in`src/base-element.js`? With auto-envelope, we only have access to a minimal set of methods.

Author: Good point, removed. It was originally included to explain that the callbacks are no longer used for those that are familiar with creating classic components:
```
All Preact-based Bento AMP extensions extend `AMP.PreactBaseElement`, which builds upon
`BaseElement` to extricate the component implementation from the aforementioned
callbacks.`
``` 

But I think without being more explicit (i.e. don't use `buildCallback`/`layoutCallback`/etc.), the sentence creates more confusion. Better to only link the resources that _should_ be used, as we do in the "AMP/Preact Bridge" and "PreactBaseElement Callbacks" sections, which it now does. :)

});
```

> The following shows the corresponding overall structure of your Preact base element implementation file (`extensions/amp-my-element/0.1/base-element.js`).

Reviewer: Should these filenames be `0.1` or `1.0`?

the default styling you're providing and allow for easy CSS classes
and/or well-structure DOM elements.

Element styles load with the element script inside an AMP document. You tell AMP which CSS or JSS belongs to this element when [registering the element](#register-element-with-AMP).

### Pre-upgrade CSS

Bento components must specify certain layout properties in order to prevent
[Cumulative Layout Shift (CLS)](https://web.dev/cls) when in use on non-AMP pages.
The following are a standard set of pre-upgrade styles to be used when no AMP
runtime or boilerplate--which typically provides this measure of stability for

> the document author--are not in use:

Reviewer: nit: Could we unwrap this paragraph into a single line for consistency?

Author: Done

];
```

> The entry for your component must have `options.wrapper = "bento"`, and may optionally include `options.npm = true` to opt the Preact component portion for automated npm publishing on AMP's release schedule.

Reviewer: ```suggestion
The entry for your component must have `options.wrapper = "bento"`. It may optionally include `options.npm = true` to publish the Preact component portion on npm automatically, following AMP's release schedule.
```

```javascript
exports.extensionBundles = [
...

>   {name: 'amp-my-element', version: '1.0', latestVersion: '0.1', options: {npm: true, wrapper: "bento"}},

Reviewer: This file is JSON now. We should format this code and the below snippet as well for length.

```json
{
  "name": "amp-my-element",
  "version": "1.0",
  "latestVersion": "0.1",
  "options": {
    "npm": true,
    "wrapper": "bento"
  }
}
```

Author: Done, thanks
